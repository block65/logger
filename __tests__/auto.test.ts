import {
  afterAll,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { createAutoConfiguredLoggerWithWaitableMock } from './helpers.js';

describe('Auto Logger', () => {
  const initialEnv = process.env;

  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterAll(() => {
    process.env = { ...initialEnv };
    jest.useRealTimers();
  });

  test('GCP', async () => {
    process.env.K_CONFIGURATION = 'cfg';
    process.env.K_SERVICE = 'svc';
    process.env.K_REVISION = 'rev';

    const [logger, callback, errBack] =
      createAutoConfiguredLoggerWithWaitableMock();

    logger.info('woo yeah!!!');
    logger.error(new TypeError('Unknown SmallInt'));
    logger.warn(new SyntaxError('this code stinks'));
    logger.warn({ err: new Error('Connection refused') });

    await expect(callback.waitUntilCalledTimes(1)).resolves.toMatchSnapshot();
    expect(errBack).toBeCalledTimes(0);
  });

  test('AWS ECS', async () => {
    process.env.ECS_AVAILABLE_LOGGING_DRIVERS = 'aws-logs';

    const [logger, callback, errBack] =
      createAutoConfiguredLoggerWithWaitableMock();

    logger.info('woo yeah!!!');
    logger.error(new Error('bad stuff'));
    logger.warn(new Error('bad stuff'));
    logger.warn({ err: new Error('bad stuff') });

    await expect(callback.waitUntilCalledTimes(1)).resolves.toMatchSnapshot();
    expect(errBack).toBeCalledTimes(0);
  });

  test('AWS Lambda', async () => {
    process.env.AWS_LAMBDA_FUNCTION_VERSION = '$LATEST';

    const [logger, callback, errBack] =
      createAutoConfiguredLoggerWithWaitableMock();

    logger.info('woo yeah!!!');
    logger.error(new Error('bad stuff'));
    logger.warn(new Error('bad stuff'));
    logger.warn({ err: new Error('bad stuff') });

    await expect(callback.waitUntilCalledTimes(1)).resolves.toMatchSnapshot();
    expect(errBack).toBeCalledTimes(0);
  });
});
