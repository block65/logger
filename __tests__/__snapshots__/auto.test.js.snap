// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auto Logger AWS ECS 1`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z		INFO	woo yeah!!!	{\\"level\\":30,\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		ERROR	bad stuff	{\\"level\\":50,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:30:22)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"},\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		WARN	bad stuff	{\\"level\\":40,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:31:21)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"},\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		WARN	bad stuff	{\\"level\\":40,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:32:28)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"},\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
]
`;

exports[`Auto Logger AWS Lambda 1`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z		INFO	woo yeah!!!	{\\"level\\":30}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		ERROR	bad stuff	{\\"level\\":50,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:41:22)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		WARN	bad stuff	{\\"level\\":40,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:42:21)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"}}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		WARN	bad stuff	{\\"level\\":40,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"bad stuff\\",\\"stack\\":\\"Error: bad stuff\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:43:28)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"}}
",
  ],
]
`;

exports[`Auto Logger GCP 1`] = `
Array [
  Array [
    "{\\"severity\\":\\"INFO\\",\\"time\\":\\"2009-02-13T23:31:30.000Z\\",\\"message\\":\\"woo yeah!!!\\",\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "{\\"severity\\":\\"ERROR\\",\\"time\\":\\"2009-02-13T23:31:30.000Z\\",\\"message\\":\\"Unknown SmallInt\\",\\"@type\\":\\"type.googleapis.com~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:19:22)\\",\\"name\\":\\"TypeError\\",\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "{\\"severity\\":\\"WARNING\\",\\"time\\":\\"2009-02-13T23:31:30.000Z\\",\\"message\\":\\"this code stinks\\",\\"err\\":{\\"name\\":\\"SyntaxError\\",\\"message\\":\\"this code stinks\\",\\"stack\\":\\"SyntaxError: this code stinks\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:20:21)\\"},\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
  Array [
    "{\\"severity\\":\\"WARNING\\",\\"time\\":\\"2009-02-13T23:31:30.000Z\\",\\"message\\":\\"Connection refused\\",\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"Connection refused\\",\\"stack\\":\\"Error: Connection refused\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/auto.test.js:21:28)\\"},\\"ctx\\":{\\"pid\\":1337,\\"hostname\\":\\"testhost\\"}}
",
  ],
]
`;
